sql/core/src/test/scala/org/apache/spark/sql/sources/PartitionedWriteSuite.scala

```diff
@@ -157,3 +161,49 @@ class PartitionedWriteSuite extends QueryTest with SharedSparkSession {
     }
   }
 }
+
+private class DetectDynamicSpeculationCommitProtocol(
+    jobId: String,
+    path: String,
+    dynamicPartitionOverwrite: Boolean)
+  extends HadoopMapReduceCommitProtocol(jobId, path, dynamicPartitionOverwrite) {
+
+  override def commitTask(taskContext: TaskAttemptContext): TaskCommitMessage = {
+    if (dynamicPartitionOverwrite) {
+      val partitionPathSet = dynamicStagingTaskFiles
+        .map(taskFile => getDynamicPartitionPath(taskFile, taskContext))
+        .map(_.toUri.getPath.stripPrefix(stagingDir.toUri.getPath +
+          Path.SEPARATOR))
+      assert(partitionPathSet.equals(partitionPaths))
+    }
+    super.commitTask(taskContext)
+  }
+}
+
+class PartitionedSpeculateWriteSuite extends QueryTest with SharedSparkSession {
+  import testImplicits._
+
+  override def sparkConf(): SparkConf = {
+    super.sparkConf
+      .set(config.SPECULATION_MULTIPLIER, 0.0)
+      .set(config.SPECULATION_QUANTILE, 0.5)
+      .set(config.SPECULATION_ENABLED, true)
+  }
+
+  test("SPARK-27194 SPARK-29302: Fix the issue that for dynamic partition overwrite, a " +
+    "task would conflict with its speculative task") {
+    withSQLConf(SQLConf.PARTITION_OVERWRITE_MODE.key -> PartitionOverwriteMode.DYNAMIC.toString,
+      SQLConf.FILE_COMMIT_PROTOCOL_CLASS.key ->
+        classOf[DetectDynamicSpeculationCommitProtocol].getName) {
+      withTable("t") {
```

Comment:
I was testing it locally, without the fix this test still passes. Can you please check once?

Reply:
It dis passed without this fix.
Because for the local file system, it would not throw exception.
But for the distributed file system, it do not allow two task write the same file.
The exception is liked `No lease on inode`.

Reply:
I have a new idea for the test:

First, we manually create a file in the format of `part-$split%05d-$jobId$ext` to simulate speculation or abortion; 
Second, we run a normal query with dynamic partition overwrite mode.

In this way, the test should be much simpler and only pass with the fix.

Reply:
will try it

Reply:
We can inherit `HadoopMapReduceCommitProtocol` and pass a constant job Id to it? e.g.

```
MyProtocol(jobId, path, dynamic) extends HadoopMapReduceCommitProtocol(0, path, dynamic)
```

Reply:
looks good, I will try, thanks a lot